flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java

```diff
@@ -288,58 +303,70 @@ private SortBuffer getBroadcastSortBuffer() throws IOException {
             return broadcastSortBuffer;
         }
 
-        broadcastSortBuffer =
-                new PartitionSortedBuffer(
-                        bufferPool,
-                        numSubpartitions,
-                        networkBufferSize,
-                        numBuffersForSort,
-                        subpartitionOrder);
+        broadcastSortBuffer = createNewSortBuffer();
         return broadcastSortBuffer;
     }
 
+    private SortBuffer createNewSortBuffer() {
+        if (numBuffersForWrite > 0) {
+            return new SortBasedPartitionSortedBuffer(
+                    bufferPool,
+                    numSubpartitions,
+                    networkBufferSize,
+                    numBuffersForSort,
+                    subpartitionOrder);
+        } else {
+            return new HashBasedPartitionSortedBuffer(
+                    bufferPool, numSubpartitions, numBuffersForSort, subpartitionOrder);
+        }
+    }
+
     private void flushSortBuffer(SortBuffer sortBuffer, boolean isBroadcast) throws IOException {
-        if (sortBuffer == null || sortBuffer.isReleased()) {
+        if (sortBuffer == null || sortBuffer.isReleased() || !sortBuffer.hasRemaining()) {
             return;
         }
-        sortBuffer.finish();
-
-        if (sortBuffer.hasRemaining()) {
-            fileWriter.startNewRegion(isBroadcast);
 
-            List<BufferWithChannel> toWrite = new ArrayList<>();
-            Queue<MemorySegment> segments = getWriteSegments();
+        Queue<MemorySegment> segments = new ArrayDeque<>(writeSegments);
+        int numBuffersToWrite =
+                numBuffersForWrite == 0
+                        ? EXPECTED_WRITE_BATCH_SIZE
+                        : Math.min(EXPECTED_WRITE_BATCH_SIZE, segments.size());
+        List<BufferWithChannel> toWrite = new ArrayList<>(numBuffersToWrite);
 
-            while (sortBuffer.hasRemaining()) {
-                if (segments.isEmpty()) {
-                    fileWriter.writeBuffers(toWrite);
-                    toWrite.clear();
-                    segments = getWriteSegments();
-                }
-
-                BufferWithChannel bufferWithChannel =
-                        sortBuffer.copyIntoSegment(checkNotNull(segments.poll()));
-                updateStatistics(bufferWithChannel.getBuffer(), isBroadcast);
-                toWrite.add(compressBufferIfPossible(bufferWithChannel));
+        fileWriter.startNewRegion(isBroadcast);
+        do {
+            if (toWrite.size() >= numBuffersToWrite) {
+                writeBuffers(toWrite);
+                segments = new ArrayDeque<>(writeSegments);
             }
 
-            fileWriter.writeBuffers(toWrite);
-        }
+            BufferWithChannel bufferWithChannel = sortBuffer.copyIntoSegment(segments.poll());
```

Comment:
It is also a bit weird that `copyIntoSegment` might pass a null segment. In consideration of the deadline I think we might rename the method to be like `getNextBuffer(@Nullable MemorySegment transitBuffer)` and add proper comments ?